### PR TITLE
Update to KFP pipelines codelab code (GH summarization)

### PR DESCRIPTION
**Pull request by amygdala**
---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/examples/638)

### CODE REVIEW

**Comment by amygdala** [comment]
This is the presubmit error, in gis-test-1172561965243961344, which AFAIK has nothing to do with my code changes:
`Exception occured; type <class 'kubernetes.client.rest.ApiException'> message`

**Comment by amygdala** [comment]
cc @zhenghuiwang , as you had kindly offered to look at the presubmit issues.

**Comment by zhenghuiwang** [comment]
Hi @amygdala 

I'm new to this example, but I dig a little into the presubmit test. The error message is 
```
gis-test-1173374306332184576:
Exception occured; type <class 'kubernetes.client.rest.ApiException'> message
```

I searched the test in the CI Kubernetes cluster and found the relevant pod [here](https://pantheon.corp.google.com/kubernetes/workload?folder=&project=kubeflow-ci&workload_list_tablesize=50&workload_list_tablequery=%255B%257B_22k_22_3A_22is_system_22_2C_22t_22_3A11_2C_22v_22_3A_22_5C_22false_5C_22_22_2C_22s_22_3Atrue%257D_2C%257B_22k_22_3A_22metadata%252Fname_22_2C_22t_22_3A10_2C_22v_22_3A_22_5C_22gis-test-1173374306332184576_5C_22_22%257D%255D&workload_list_tablesort=-status%252Fstatus_sort_key%252Cmetadata%252Fname)

The strange thing is that the pod succeeded. Looking through the [logs](https://pantheon.corp.google.com/logs/viewer?interval=NO_LIMIT&project=kubeflow-ci&folder&minLogLevel=0&expandAll=false&timestamp=2019-09-16T21:34:08.373000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22kubeflow-ci%22%0Aresource.labels.location%3D%22us-east1-b%22%0Aresource.labels.cluster_name%3D%22kf-vmaster-n00%22%0Aresource.labels.namespace_name%3D%22kubeflow%22%0Aresource.labels.pod_name%3D%22gis-test-1173374306332184576-master-0%22%0Aresource.labels.container_name%3D%22tensorflow%22), I think the relevant test is the [testing/tfjob.py](https://github.com/kubeflow/examples/blob/master/github_issue_summarization/testing/tfjob_test.py#L59)

Maybe you can try to verify if this test is the problem by manually run it (there are some comments on how to do it in the test) or run the presubmit tests without this file(delete it and run tests)?

We can ask other folks if that test is the cause.

**Comment by amygdala** [comment]
Thank you!  [Assuming that's the culprit, that test is unrelated to this particular PR, as the pipeline doesn't use tfjob].

Jeremy, it looks like you might have been the original author of testing/tfjob.py? I wonder if the error is ksonnet-related.  @jlewi

**Comment by zhenghuiwang** [comment]
I created an empty PR to rerun the presubmit test on master. 

The testing argo workflow shows the `tfjob-test` step succeeded but there are errors in the logs. Hope it helps debug. I added retry with `@retry(stop_max_attempt_number=5, wait_fixed=5000)` but it does fix it.

```
...

INFO|2019-09-16T23:43:52|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/testing/py/kubeflow/testing/util.py|41| Running: ks-13 apply test-env-e084 -c tfjob
cwd=/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/examples/github_issue_summarization/ks_app
INFO|2019-09-16T23:43:52|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/testing/py/kubeflow/testing/util.py|56| Subprocess output:
INFO|2019-09-16T23:44:07|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/testing/py/kubeflow/testing/util.py|62| level=info msg="Applying tfjobs kubeflow.gis-test-1173742360962535424"
INFO|2019-09-16T23:44:07|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/testing/py/kubeflow/testing/util.py|62| level=info msg="Creating non-existent tfjobs kubeflow.gis-test-1173742360962535424"
INFO|2019-09-16T23:44:09|tfjob_test.py|71| Created job gis-test-1173742360962535424 in namespaces kubeflow
INFO|2019-09-16T23:44:09|tfjob_test.py|74| Waiting for job to finish.
ERROR|2019-09-16T23:44:09|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/tf-operator/py/kubeflow/tf_operator/tf_job_client.py|155| There was a problem waiting for Job gis-test-1173742360962535424.gis-test-117
3742360962535424; Exception; (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Date': 'Mon, 16 Sep 2019 23:44:09 GMT', 'Audit-Id': '0109f5ad-86b0-47d1-97cd-36b93fc34b4d', 'Content-Length': '19', 'Content-Type': 'text/plain; charset=utf-8', 'X-Content-Type-Options': 'nosniff'})
HTTP response body: 404 page not found
ERROR|2019-09-16T23:44:09|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/tf-operator/py/kubeflow/tf_operator/test_runner.py|72| There was a problem running the job; Exception (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Date': 'Mon, 16 Sep 2019 23:44:09 GMT', 'Audit-Id': '0109f5ad-86b0-47d1-97cd-36b93fc34b4d', 'Content-Length': '19', 'Content-Type': 'text/plain; charset=utf-8', 'X-Content-Type-Options': 'nosniff'})
HTTP response body: 404 page not found
Traceback (most recent call last):
 File "/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/tf-operator/py/kubeflow/tf_operator/test_runner.py", line 53, in run_test
   test_func()
 File "tfjob_test.py", line 79, in test_train
   status_callback=tf_job_client.log_status)
 File "/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/tf-operator/py/kubeflow/tf_operator/tf_job_client.py", line 208, in wait_for_job
   status_callback=status_callback)
 File "/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/tf-operator/py/kubeflow/tf_operator/tf_job_client.py", line 150, in wait_for_condition
   results = thread.get(TIMEOUT)
 File "/usr/lib/python2.7/multiprocessing/pool.py", line 567, in get
   raise self._value
ApiException: (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Date': 'Mon, 16 Sep 2019 23:44:09 GMT', 'Audit-Id': '0109f5ad-86b0-47d1-97cd-36b93fc34b4d', 'Content-Length': '19', 'Content-Type': 'text/plain; charset=utf-8', 'X-Content-Type-Options': 'nosniff'})
HTTP response body: 404 page not found
INFO|2019-09-16T23:44:09|/mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/src/kubeflow/testing/py/kubeflow/testing/test_util.py|151| Creating /mnt/test-data-volume/kubeflow-examples-presubmit-gis-640-c0377ce-5424-7c8f/outpu
t/artifacts/junit_test_train.xml
Exception RuntimeError: RuntimeError('cannot join current thread',) in <bound method ApiClient.__del__ of <kubernetes.client.api_client.ApiClient object at 0x7f23370aee50>> ignored
```

**Comment by jlewi** [comment]
The test is running on cluster kf-master-n00.  The TFJob CRD in that cluster is

```
kubectl -n kubeflow get crd -o yaml tfjobs.kubeflow.org
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apiextensions.k8s.io/v1beta1","kind":"CustomResourceDefinition","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"tfjob","app.kubernetes.io/instance":"tf-job-operator","app.kubernetes.io/managed-by":"kfctl","app.kubernetes.io/name":"tf-job-operator","app.kubernetes.io/part-of":"kubeflow","app.kubernetes.io/version":"v0.6","kustomize.component":"tf-job-operator"},"name":"tfjobs.kubeflow.org"},"spec":{"additionalPrinterColumns":[{"JSONPath":".status.conditions[-1:].type","name":"State","type":"string"},{"JSONPath":".metadata.creationTimestamp","name":"Age","type":"date"}],"group":"kubeflow.org","names":{"kind":"TFJob","plural":"tfjobs","singular":"tfjob"},"scope":"Namespaced","subresources":{"status":{}},"validation":{"openAPIV3Schema":{"properties":{"spec":{"properties":{"tfReplicaSpecs":{"properties":{"Chief":{"properties":{"replicas":{"maximum":1,"minimum":1,"type":"integer"}}},"PS":{"properties":{"replicas":{"minimum":1,"type":"integer"}}},"Worker":{"properties":{"replicas":{"minimum":1,"type":"integer"}}}}}}}}}},"version":"v1","versions":[{"name":"v1","served":true,"storage":true},{"name":"v1beta2","served":true,"storage":false}]}}
  creationTimestamp: 2019-09-17T12:17:35Z
  generation: 1
  labels:
    app.kubernetes.io/component: tfjob
    app.kubernetes.io/instance: tf-job-operator
    app.kubernetes.io/managed-by: kfctl
    app.kubernetes.io/name: tf-job-operator
    app.kubernetes.io/part-of: kubeflow
    app.kubernetes.io/version: v0.6
    kustomize.component: tf-job-operator
  name: tfjobs.kubeflow.org
  resourceVersion: "2086"
  selfLink: /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/tfjobs.kubeflow.org
  uid: 21bd216b-d945-11e9-8eeb-42010af000b3
spec:
  additionalPrinterColumns:
  - JSONPath: .status.conditions[-1:].type
    name: State
    type: string
  - JSONPath: .metadata.creationTimestamp
    name: Age
    type: date
  group: kubeflow.org
  names:
    kind: TFJob
    listKind: TFJobList
    plural: tfjobs
    singular: tfjob
  scope: Namespaced
  subresources:
    status: {}
  validation:
    openAPIV3Schema:
      properties:
        spec:
          properties:
            tfReplicaSpecs:
              properties:
                Chief:
                  properties:
                    replicas:
                      maximum: 1
                      minimum: 1
                      type: integer
                PS:
                  properties:
                    replicas:
                      minimum: 1
                      type: integer
                Worker:
                  properties:
                    replicas:
                      minimum: 1
                      type: integer
  version: v1
  versions:
  - name: v1
    served: true
    storage: true
  - name: v1beta2
    served: true
    storage: false
status:
  acceptedNames:
    kind: TFJob
    listKind: TFJobList
    plural: tfjobs
    singular: tfjob
  conditions:
  - lastTransitionTime: 2019-09-17T12:17:35Z
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: null
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  storedVersions:
  - v1
```

The job however is using v1beta1
https://github.com/kubeflow/examples/blob/78a79e72dc809cc01e0f6a3af17697805ca3d30a/github_issue_summarization/ks_app/components/tfjob.jsonnet#L10

So the job is using v1beta1 but only v1beta2 and v1 are registered. So I suspect the TFJob is not actually created. As a result, when we issue a get for the TFJob it returns a 404 because the job doesn't exist.

You could try fixing the test by setting the version to v1 here
https://github.com/kubeflow/examples/blob/78a79e72dc809cc01e0f6a3af17697805ca3d30a/github_issue_summarization/ks_app/components/tfjob.jsonnet#L10

If that doesn't work you could mark the tfjob_test.py with an expected failure annotation to skip it
https://docs.python.org/3/library/unittest.html#unittest.expectedFailure

**Comment by amygdala** [comment]
Hmm, adding the unittest annotation gives this error, which I'm not clear how to dig in to further (no info in the build log): `Exception occured; type <class 'unittest.case._ExpectedFailure'> message`. Might try just commenting out what could be the culprit test.

**Comment by texasmichelle** [comment]
Instead of commenting out the problematic test, try replacing `unittest` with `test_util`, e.g.:
`@test_util.expectedFailure`

**Comment by amygdala** [comment]
Hmm, using `@test_util.expectedFailure` didn't seem to have the desired effect :(. 
The build log doesn't give much info though.

**Comment by texasmichelle** [comment]
Clearly that was the wrong advice, since I found this in the logs:
```
$ kubectl -n kubeflow-test-infra logs kubeflow-examples-presubmit-gis-638-046171f-1667-62c9-697619872 main
Traceback (most recent call last):
  File "tfjob_test.py", line 38, in <module>
    class TFJobTest(test_util.TestCase):
  File "tfjob_test.py", line 55, in TFJobTest
    @test_util.expectedFailure
AttributeError: 'module' object has no attribute 'expectedFailure'
```
Since that didn't work, I'm fine with commenting out the problematic test and addressing in a follow-on PR.

Viewing the detailed logs is a pain, but here's how I found them:
1. Connect to the `kubeflow-ci` project, view pods in the `kubeflow-test-infra` namespace
1. Locate the identifier unique to the PR by clicking on Details next to the presubmit check failure in GitHub. In the Prow dashboard, under Artifacts, look at the string appended to the filename (in this case `kubeflow-examples-presubmit-gis-638-046171f-1667-62c9.yaml`)
1. Grep for the first few characters of the unique ID to find related pods. View logs for the most recent pod in error state.

File: github_issue_summarization/pipelines/README.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 This Kubeflow Pipelines example shows how to build a web app that summarizes GitHub issues using Kubeflow Pipelines to train and serve a model.
 The pipeline trains a [Tensor2Tensor](https://github.com/tensorflow/tensor2tensor/) model on GitHub issue data, learning to predict issue titles from issue bodies. It then exports the trained model and deploys the exported model using  [Tensorflow Serving](https://github.com/tensorflow/serving). The final step in the pipeline launches a web app, which interacts with the TF-Serving instance in order to get model predictions.
 
-You can follow this example as a codelab: [g.co/codelabs/kubecon18](https://g.co/codelabs/kubecon18).    
-Or, you can run it as a [Cloud shell Tutorial](https://console.cloud.google.com/?cloudshell=true&cloudshell_git_repo=https://github.com/kubeflow/examples&working_dir=github_issue_summarization/pipelines&cloudshell_tutorial=tutorial.md).  The source for the Cloud Shell tutorial is [here](tutorial.md).
+You can follow this example as a codelab: [g.co/codelabs/kubecon18](https://g.co/codelabs/kubecon18).
```
Comment (texasmichelle) [inline github_issue_summarization/pipelines/README.md]: Let's use this shortlink instead: `g.co/codelabs/kfp-gis`

Comment (amygdala) [reply github_issue_summarization/pipelines/README.md]: done, thanks for the catch.

**Comment by amygdala** [comment]
Thank you!  and thanks for the info on looking at the CI logs.

**Pull request merged**